### PR TITLE
Add audmath.distance()

### PR DESCRIPTION
**Pull request by hagenw**
This adds a function to compute a pairwise Euclidean or cosine distance between vectors (points, embeddings, ...).
This invents nothing new, but provides a solution that doesn't require `scipy`.

image

### CODE REVIEW

**Comment by hagenw** [comment]
@schruefer is this in line how you normally calculate distances between vectors?

File: tests/test_distance.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+                [
+                    scipy.spatial.distance.cosine(
+                        x1_expected[m, :],
+                        x2_expected[n, :],
+                    )
+                    for n in range(x2_expected.shape[0])
+                ]
+            )
+        expected = np.array(expected)
+    expected = expected.squeeze()
```
Comment (schruefer) [inline tests/test_distance.py:102]: expected might not be assigned if no valid metric is selected

File: audmath/core/api.py [hunk 2]
```diff
@@ -98,6 +98,73 @@ def db(
     return x
 
 
+def distance(
+    u: typing.Union[typing.List, typing.Tuple, np.ndarray],
+    v: typing.Union[typing.List, typing.Tuple, np.ndarray],
+    *,
+    metric: str = 'cosine',
```
Comment (schruefer) [inline audmath/core/api.py]: I would use 'euclidean' as default metric, as I think that it is more general and more often used than cosine distance.

Even if we would later use the cosine distance for similarity measurement.

**Comment by schruefer** [comment]
It might be also useful to add the manhattan distance  as an additional metric.

Looks good to me!

Even though I have no idea how we should handle different edge cases/errors.

e.g. 
>>> from scipy.spatial import distance
>>> distance.cosine([0],[1])
0

vs

>>> audmath.distance([0,0], [1,1])
nan

or 

>>> distance.cosine(2,8)
ValueError: Input vector should be 1-D.

vs

>>> audmath.distance(1, 2)
0.0

**Comment by hagenw** [comment]
> This doesn't look like the right formula to me :)
> 
> image

image

and

```math
\lVert u - v \rVert_2 = \sqrt{\sum_k |u_k - v_k|^2} = \sqrt{\sum_k (u_k - v_k)^2} = \sqrt{(u - v) (u-v)} = \lVert u-v \lVert
```

Or are you referring to my usage of `u` and `v` as variables for the arrays and for the vectors, which might be indeed not ideal.

**Comment by hagenw** [comment]
Edge cases are a valid point, I will have a closer look at them.

For cases that are not covered by the type hint like `audmath.distance(1, 2)` which does not provide an array, list or tuple, we usually don't check what the outcome might be. But maybe we should change that here?

File: audmath/core/api.py (hunk 2 above)
Comment (hagenw) [reply audmath/core/api.py]: So you would select `euclidean` as default for `distance()` and `cosine` for `similarity()` as these are the commonly used ones, correct?

Comment (schruefer) [reply audmath/core/api.py]: Yes, correctly, I don't have a strong opinion to use cosine similarity, but i would say this is the most practical choice.

**Comment by frankenjoe** [comment]
> Or are you referring to my usage of `u` and `v` as variables for the arrays and for the vectors, which might be indeed not ideal.

In the first place I was wondering about the lower 2 there.

**Comment by hagenw** [comment]
When calculating the pairwise distance for the `euclidean` distances we might run out of memory with the current fast calculation approach. This is covered by the `threshold` option in [scipy.spatial.distance_matrix](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance_matrix.html).

As calculating the distance and similarity is not a core thing we need to do, I wonder if we simple should stick with the available solutions from `scipy`?

There we have:

* [scipy.spatial.distance.cosine](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.cosine.html)
* [scipy.spatial.distance.euclidean](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.euclidean.html#scipy.spatial.distance.euclidean)
* [scipy.spatial.distance_matrix](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance_matrix.html) for p-norms
* [scipy.spatial.distance.pdist](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.pdist.html#scipy.spatial.distance.pdist) for calculating distance matrices for a lot more norms

**Comment by schruefer** [comment]
> As calculating the distance and similarity is not a core thing we need to do, I wonder if we simple should stick with the available solutions from scipy when ever needed and not add anything to audmath?

I am fine with sticking to the available solutions from scipy.

**Comment by hagenw** [comment]
OK, I will then close for now.
If we change our opinion later on we can of cause still add it.

**Pull request closed without merging**